src/Authentication/JWTAuthenticator.php

```diff
@@ -25,201 +34,252 @@
     }
 
     /**
-     * @param array $data
-     * @param HTTPRequest $request
+     * @param array                 $data
+     * @param HTTPRequest           $request
      * @param ValidationResult|null $result
      * @return Member|null
      * @throws OutOfBoundsException
      * @throws BadMethodCallException
+     * @throws Exception
      */
-    public function authenticate(array $data, HTTPRequest $request, ValidationResult &$result = null)
+    public function authenticate(array $data, HTTPRequest $request, ValidationResult &$result = null): ?Member
     {
         if (!$result) {
             $result = new ValidationResult();
         }
         $token = $data['token'];
 
-        return $this->validateToken($token, $request, $result);
+        /** @var JWTRecord $record */
+        list($record, $status) = $this->validateToken($token, $request);
+
+        // Report success!
+        if ($status === TokenStatusEnum::STATUS_OK) {
+            return $record->Member();
+        }
+
+        // Add errors to result
+        $result->addError(
+            $this->getErrorMessage($status),
+            ValidationResult::TYPE_ERROR,
+            $status
+        );
+        return null;
     }
 
     /**
-     * @param Member $member
+     * Generate a new JWT token for a given request, and optional (if anonymous_allowed) user
+     *
+     * @param HTTPRequest            $request
+     * @param Member|MemberExtension $member
      * @return Token
      * @throws ValidationException
-     * @throws BadMethodCallException
      */
-    public function generateToken(Member $member)
+    public function generateToken(HTTPRequest $request, Member $member): Token
     {
-        $this->setKeys();
         $config = static::config();
-        $uniqueID = uniqid(Environment::getEnv('JWT_PREFIX'), true);
+        $uniqueID = uniqid($this->getEnv('JWT_PREFIX', ''), true);
 
-        $request = Controller::curr()->getRequest();
-        $audience = $request->getHeader('Origin');
+        // Create new record
+        $record = new JWTRecord();
+        $record->UID = $uniqueID;
+        $record->UserAgent = $request->getHeader('User-Agent');
+        $member->AuthTokens()->add($record);
+        if (!$record->isInDB()) {
+            $record->write();
+        }
 
+        // Create builder for this record
         $builder = new Builder();
+        $now = DBDatetime::now()->getTimestamp();
         $token = $builder
             // Configures the issuer (iss claim)
-            ->setIssuer($audience)
+            ->setIssuer($request->getHeader('Origin'))
             // Configures the audience (aud claim)
             ->setAudience(Director::absoluteBaseURL())
             // Configures the id (jti claim), replicating as a header item
             ->setId($uniqueID, true)
             // Configures the time that the token was issue (iat claim)
-            ->setIssuedAt(time())
+            ->setIssuedAt($now)
```

Comment:
This should be configurable actually, neither my original time() nor $now is correct to me, it should take in to account any options to set it in the future from config I'd think

Reply:
When do we want to set a different issued at date? Maybe another discussion huh.

Reply:
Yeah, probably better off as a separate issue

Reply:
Issue: https://github.com/Firesphere/silverstripe-graphql-jwt/issues/25